device/alpaka/include/traccc/alpaka/utils/definitions.hpp

```diff
@@ -11,15 +11,42 @@
 #include <alpaka/example/ExampleDefaultAcc.hpp>
 
 #ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
-#include "vecmem/utils/cuda/copy.h"
+#include <vecmem/utils/cuda/copy.hpp>
 #endif
 
-#include "vecmem/utils/copy.hpp"
+#include <vecmem/utils/copy.hpp>
 
 namespace traccc::alpaka {
 
+#ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
+#define WARP_SIZE 32
+#else
+#define WARP_SIZE 4
+#endif
+
 using Dim = ::alpaka::DimInt<1>;
 using Idx = uint32_t;
 using WorkDiv = ::alpaka::WorkDivMembers<Dim, Idx>;
 
+using Acc = ::alpaka::ExampleDefaultAcc<Dim, Idx>;
+using Host = ::alpaka::DevCpu;
+using Queue = ::alpaka::Queue<Acc, ::alpaka::NonBlocking>;
+
+template <typename TAcc>
+inline WorkDiv makeWorkDiv(Idx blocksPerGrid,
+                           Idx threadsPerBlockOrElementsPerThread) {
+#ifdef ALPAKA_ACC_GPU_CUDA_ENABLED
+    if constexpr (std::is_same_v<TAcc, ::alpaka::AccGpuCudaRt<Dim, Idx>>) {
+        const auto elementsPerThread = Idx{1};
+        return WorkDiv{blocksPerGrid, threadsPerBlockOrElementsPerThread,
+                       elementsPerThread};
+    } else
+#endif
```

Comment:
Interesting. Is this piece of code non-functional when CUDA is not being used? I would've hoped that the Alpaka code around CUDA could still be referred to, even if CUDA was "not active". :thinking: Since the whole point of Alpaka in my mind would be to do as little pre-processor magic on our side as possible.

Reply:
This is certainly something myself and Stewart are talking about, as this situation gets worse when you start adding in `if CUDA if HIP else` etc...

https://github.com/cms-sw/cmssw/blob/70c7fb8ed2ff22abb8a635c438fa7dd2313ebb3b/HeterogeneousCore/AlpakaInterface/interface/workdivision.h#L21-L76

Gives an idea of how they deal with this in CMSSW, but I think the alpaka recommended way does use some small amount of pre-processor defines, when you get into code for a specific accelerator, at least if you want to use code that can only work if you are using the compiler for that accelerator.

Examples on the alpaka side for that:

Define-based : https://github.com/alpaka-group/alpaka/blob/develop/example/kernelSpecialization/src/kernelSpecialization.cpp

Tag Based : https://github.com/alpaka-group/alpaka/blob/develop/example/tagSpecialization/src/tagSpecialization.cpp